drivers/ethernet/eth_stm32_hal.c

```diff
@@ -932,3 +1130,285 @@ static struct eth_stm32_hal_dev_data eth0_data = {
+	}
+
+	/* Save new ratio */
+	eth_dev_data->clk_ratio_adj = ratio;
+
+	/* Update addend register */
+	addend_val = UINT32_MAX * eth_dev_data->clk_ratio * ratio;
+
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	heth->Instance->MACTSAR = addend_val;
+	heth->Instance->MACTSCR |= ETH_MACTSCR_TSADDREG;
+	while (heth->Instance->MACTSCR & ETH_MACTSCR_TSADDREG_Msk) {
+		/* spin lock */
+	}
+#else
+	heth->Instance->PTPTSAR = addend_val;
+	heth->Instance->PTPTSCR |= ETH_PTPTSCR_TSARU;
+	while (heth->Instance->PTPTSCR & ETH_PTPTSCR_TSARU_Msk) {
+		/* spin lock */
+	}
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
+	ret = 0;
+
+error:
+	irq_unlock(key);
+
+	return ret;
+}
+
+static const struct ptp_clock_driver_api api = {
+	.set = ptp_clock_stm32_set,
+	.get = ptp_clock_stm32_get,
+	.adjust = ptp_clock_stm32_adjust,
+	.rate_adjust = ptp_clock_stm32_rate_adjust,
+};
+
+static int ptp_stm32_init(const struct device *port)
+{
+	const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(mac));
+	struct eth_stm32_hal_dev_data *eth_dev_data = DEV_DATA(dev);
+	const struct eth_stm32_hal_dev_cfg *eth_cfg = DEV_CFG(dev);
+	struct ptp_context *ptp_context = port->data;
+	ETH_HandleTypeDef *heth = &eth_dev_data->heth;
+	int ret;
+	uint32_t ptp_hclk_rate;
+	uint32_t ss_incr_ns;
+	uint32_t addend_val;
+
+	eth_dev_data->ptp_clock = port;
+	ptp_context->eth_dev_data = eth_dev_data;
+
+	/* Mask the Timestamp Trigger interrupt */
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	heth->Instance->MACIER &= ~(ETH_MACIER_TSIE);
+#else
+	heth->Instance->MACIMR &= ~(ETH_MACIMR_TSTIM);
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
+	/* Enable timestamping */
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	heth->Instance->MACTSCR |= ETH_MACTSCR_TSENA;
+#else
+	heth->Instance->PTPTSCR |= ETH_PTPTSCR_TSE;
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
+	/* Query ethernet clock rate */
+	ret = clock_control_get_rate(eth_dev_data->clock,
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+		(clock_control_subsys_t *)&eth_cfg->pclken,
+#else
+		(clock_control_subsys_t *)&eth_cfg->pclken_ptp,
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+		&ptp_hclk_rate);
+	if (ret) {
+		LOG_ERR("Failed to query ethernet clock");
+		return -EIO;
+	}
+
+	/* Program the subsecond increment register based on the PTP clock freq */
+	if (NSEC_PER_SEC % CONFIG_ETH_STM32_HAL_PTP_CLOCK_SRC_HZ != 0) {
+		LOG_ERR("PTP clock period must be an integer nanosecond value");
+		return -EINVAL;
+	}
+	ss_incr_ns = NSEC_PER_SEC / CONFIG_ETH_STM32_HAL_PTP_CLOCK_SRC_HZ;
+	if (ss_incr_ns > UINT8_MAX) {
+		LOG_ERR("PTP clock period is more than %d nanoseconds", UINT8_MAX);
+		return -EINVAL;
+	}
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	heth->Instance->MACSSIR = ss_incr_ns << ETH_MACMACSSIR_SSINC_Pos;
+#else
+	heth->Instance->PTPSSIR = ss_incr_ns;
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
+	/* Program timestamp addend register */
+	eth_dev_data->clk_ratio =
+		(double)CONFIG_ETH_STM32_HAL_PTP_CLOCK_SRC_HZ / ptp_hclk_rate;
+	eth_dev_data->clk_ratio_adj = 1.0;
```

Comment:
Why is it 1.0?

Reply:
We determine `clk_ratio` as ratio between desired PTP frequency and HCLK frequency. Now, we might need to correct clock because of crystal misspec/drift - and that's what clk_ratio_adj is for. It starts as 1.0 and is allowed to be in bounds (0.9, 1.1).

Reply:
Please use define with meaningful name or put short comment Witz description here.

Reply:
Added a comment - please let me know if it doesn't explain it sufficiently.